### PR TITLE
Allow stats hooks to be registered for ringpop extensions

### PR DESCRIPTION
**Pull request by jwolski**
@jmccarthy14 @Raynos This'll allow ringpop serial to return stats over the ringpop /admin/stats endpoint

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -44,12 +44,31 @@ var HOST_PORT_PATTERN = /^(\d+.\d+.\d+.\d+):\d+$/;
 var MAX_JOIN_DURATION = 300000;
 var PROXY_REQ_PROPS = ['keys', 'dest', 'req', 'res'];
 
+var ArgumentRequiredError = TypedError({
```
Comment (Raynos) [inline index.js]: suggestion: consider placing these errors in an errors.js file

This makes all the errors visible in a single file, might be more friendly.

Comment (jwolski) [inline index.js]: yep, agreed. i feel we're getting to that point where it'd be useful to have such a file.

File: index.js [hunk 2]
```diff
@@ -813,6 +860,26 @@ RingPop.prototype.proxyReq = function proxyReq(opts) {
     this.requestProxy.proxyReq(opts);
 };
 
+RingPop.prototype.registerStatsHook = function registerStatsHook(hook) {
+    if (!hook) {
+        throw ArgumentRequiredError({ argument: 'hook' });
+    }
+
+    if (!hook.name) {
+        throw FieldRequiredError({ argument: 'hok', field: 'name' });
```
Comment (Raynos) [inline index.js]: typo: 'hook'

File: index.js [hunk 3]
```diff
@@ -813,6 +860,26 @@ RingPop.prototype.proxyReq = function proxyReq(opts) {
...
+    }
+
+    if (typeof hook.getStats !== 'function') {
+        throw MethodRequiredError({ argument: 'hook', method: 'getStats' });
+    }
+
+    if (this.statsHooks[hook.name]) {
```
Comment (Raynos) [inline index.js]: consider using the `.isStatsHookRegistered()` method.

File: test/index_test.js [hunk 4]
```diff
@@ -190,3 +190,89 @@ test('no opts does not break handleOrProxy', function t(assert) {
...
+    ringpop.registerStatsHook({
+        name: 'myhook',
+        getStats: function getIt() {
+            return {
+                numQueues: 10
+            };
+        }
+    });
+
+    assert.ok(ringpop.getStats().hooks.myhook, 'has hook stats');
```
Comment (Raynos) [inline test/index_test.js]: suggestion: add an assertion to see that `myhook` deep equals { numQueues: 10 }

File: index.js (hunk 3 above)
Comment (jwolski) [inline index.js]: lovely idea

File: test/index_test.js (hunk 4 above)
Comment (jwolski) [inline test/index_test.js]: legit

**Pull request merged**